core/server/api/README.md

```diff
@@ -20,9 +20,92 @@ Each request goes through the following stages:
 The framework we are building pipes a request through these stages depending on the API controller implementation.
 
 
+## Options
+
+Is a class. We pass this instance per reference. The target function can modify the original instance.
+No need to return the class instance.
+
+### Structure
+
+```
+{
+  apiOptions: Object,
+  modelOptions: Object,
+  data: Object,
+  queryData: Object
+}
+```
+
+### Example
+
+```
+{
+  apiOptions: {
+    include: 'tags'
+  },
+  modelOptions: {
+    withRelated: ['tags']
+  },
+  data: {
+    posts: []
+  },
+  queryData: {
```

Comment:
I could not get rid of the inconsistenty of sometimes having query params in model options and sometimes in query data, this is because the model layer is inconsistent. I def need to think about this again before merging.

Reply:
It would be nice to have consistent function signatures for the model layer. but also we don't want to rewrite the whole model layer now and break v0.1 or have to duplicate it and start fresh. Maybe we could wrap the models for now with consistent methods?
```js
const wrap = Model => Model.extend({}, {
  edit(data, options) {
    return Model.edit(
      Object.assign({}, data, {id: undefined}),
      Object.assign({}, options, id: data.id})
    );
  },
  destroy(data, options) {
    return Model.destroy(Object.assign({}, options, {id: data.id}));
  }
});
```
Then in the controller definition we can do something like
```js
// core/server/api/v2/thing.js
module.exports = {
  docName: 'thing',
  modelName: 'Thing',
  validations: {/* blah */},
  query(options, Model) {
    // model here is  wrap(models[modelName]) - done by the framework
  }
}
```

That way we can have whatever nice model API we want now, and then implement it properly later

Reply:
Good input 👍 
Something like this could work to not having to rely on the model inconsistency.
I need to figure out which model actions use a mixture of options & data and why they are doing it.

> return Model.destroy(Object.assign({}, options, {id: data.id}));

Is this correct?

Shouldn't it be

> return Model.destroy(options.queryData, options.modelOptions)

Reply:
Danke :+1: yeah i think it would be nice and hopefully not too much overhead.

> Shouldn't it be
>> `return Model.destroy(options.queryData, options.modelOptions)`

That's how we would call it in the controller, but the wrapper layer would have to merge it into one object because the `destroy` method currently only takes options - does that make sense?

Reply:
1st problem i see: e.g. user model uses api lot's of more actions than add,edit,findOne,findPage, destroy.
That means the wrapper needs to wrap much more fns.

Reply:
> That's how we would call it in the controller, but the wrapper layer would have to merge it into one object because the destroy method currently only takes options - does that make sense?

Ah okay, yes. I thought this was the controller. Sry

Reply:
I went through the query data we use for model options:

findPage
- status: is used to add extra where statements. can def live in queryData later

destroy
- id: target id to destroy. can definitely live in queryData. `destroy({id: id}, modelOptions)`

edit:
- id: to differentiate between data to update and the actual target model id. here we would need to accept that the mode layer will later do `edit(data, options) ... omit(options, 'id')` i think?

add:
- id: probably used to add a model with a specific id - would be the same as edit

I think that is doable, the only tricky thing is the edit/add action with id property?

Reply:
The idea is in general nice, but we don't add the idea of a model wrapper for now.
The API controller can define options and data. And if the model changes, the API controller can simply adapt the allowed options and data.